src/block-data-providers/base-block-data-provider.abstract.ts

```diff
@@ -45,4 +49,30 @@ export abstract class BaseBlockDataProvider<OperationState> {
             state,
         );
     }
+
+    abstract getBlockHash(height: number): Promise<string>;
+
+    async traceReorg(): Promise<number> {
+        let state = await this.blockStateService.getCurrentBlockState();
+
+        if (state == null) {
+            return null;
+        }
+
+        while (true) {
```

Comment:
I don't like `while true` loops because they could lead to nasty infinite loop bugs. We should replace it with a suitable condition.